### PR TITLE
Improve lagged ensemble performance and versatility

### PR DESCRIPTION
**Pull request by nbren12**
# Earth-2 MIP Pull Request

## Description

I have been trying to score graphcast, but because it is big, I was first saving the forecasts to hindcast directory using earht2mip.time_collection instead of scoring it online with earth2mip.lagged_ensembles. Unfortunately, there is some bug such that at the lead time =0 forecast is not the same as the initial condition. 

This error is not present in the lagged ensemble script so I would like to use it to score graphcast directly. The lagged ensemble script currently does not handle the grid information properly and the scoring is very slow. This PR addresses both issues.

I add a memory and compute efficient implementation of CRPS in torch. The code relies on xskillscore/properscoring, but when run on the GPU this used the O(n^2) kernel method and runs out of memory when there are many channels. I did not use the modulus version of CRPS because it lacks an efficient yet exact implementation of CRPS. The kernel method is inefficient and the binned method is inexact. A bonus is that I can remove the xskillscores/properscoring dep. This code could be upstreamed to modulus core.

## Checklist

- [ ] I am familiar with the Contributing Guidelines.
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The CHANGELOG.md is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2mip/issues) is linked to this pull request.

### CODE REVIEW

File: earth2mip/forecasts.py [hunk 1]
```diff
@@ -70,6 +70,25 @@ async def __getitem__(self, i: int):
...
+
+    @property
+    def grid(self):
+        return self.forecast.grid
+    
+    async def __getitem__(self, i: int):
+        async for x in self.forecast[i]:
+            yield x[:, self._index]
+
+
```
Comment (dallasfoster) [inline earth2mip/forecasts.py:91]: I don't understand the motivation for making this an object, or an object of type `Forecast`. What is a `Forecast` and why should a function that selects channels be a type of it? This seems like it should be more of a utility function rather than a class.

File: earth2mip/lagged_ensembles/__main__.py
Comment (dallasfoster) [inline earth2mip/lagged_ensembles/__main__.py:1]: This is a lot of code to be placed in a file so non-descript as `__main__.py`. Can we consider renaming and moving some code for clarity, similar to what you did for `score.py`?

File: earth2mip/lagged_ensembles/core.py [hunk 2]
```diff
@@ -138,26 +124,107 @@ async def yield_lagged_ensembles(
...
+                elif rank == dst:
+                    torch.distributed.recv(recv_buffer, src=src)
+                    buffers[src] = recv_buffer.clone()
+    else:
+        buffers = {rank: y}
 
-    This is one way to implement the done criteria which is less easily
-    parallelized. I am leaving it in the code for educational value only.
+    return buffers  # noqa
+
```
Comment (dallasfoster) [inline earth2mip/lagged_ensembles/core.py:202]: This code seems to be misplaced in this file. A lot of the functionality here does not seem to rely on the assumption of `lagged ensemble`. The design principle should be to move these utilities elsewhere, in a distributed utliity focused area (maybe even modulus.distributed) and to generalize sufficiently so that we can reuse code if necessary.

File: earth2mip/lagged_ensembles/score.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import torch
+
+import earth2mip.grid
+from earth2mip.crps import crps_from_empirical_cdf
+
+logger = logging.getLogger(__name__)
+
+
+def weighted_average(x, w, dim):
+    return torch.mean(x * w, dim) / torch.mean(w, dim)
```
Comment (dallasfoster) [inline earth2mip/lagged_ensembles/score.py:12]: Why does this function exist here? Should this not exist elsewhere? Modulus has weighted average utilities.

File: earth2mip/lagged_ensembles/score.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def area_average(grid: earth2mip.grid.LatLonGrid, x: torch.Tensor):
+    lat = torch.tensor(grid.lat, device=x.device)[:, None]
+    cos_lat = torch.cos(torch.deg2rad(lat))
+    return weighted_average(x, cos_lat, dim=[-2, -1])
```
Comment (dallasfoster) [inline earth2mip/lagged_ensembles/score.py:18]: Why is this here? Modulus has lat and global averaging utilities.

File: earth2mip/lagged_ensembles/score.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def score(
```
Comment (dallasfoster) [inline earth2mip/lagged_ensembles/score.py:21]: Does this function need to assume the use of lagged ensembles? Is there anyway to make a more general `score` function and one way to use it would be to plug in lagged ensembles? Again, I think we should deisolate much of the code here to be more broadly usable.

File: earth2mip/lagged_ensembles/score.py (hunk 3 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/score.py:12]: I didn't know that modulus had weighted averaged utilities. It's one line of code, and only used as a helper here. Not public API.

File: earth2mip/lagged_ensembles/score.py (hunk 4 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/score.py:18]: See reply here: https://github.com/NVIDIA/earth2mip/pull/154#discussion_r1440820519

File: earth2mip/lagged_ensembles/score.py (hunk 5 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/score.py:21]: It's specific to lagged ensembles in that the m=0 lag is considered the deterministic skill.

But, this is a change of the implementation of scoring for the lagged ensemble script, and not public API. Perhaps it could be generalized elsewhere, but I think it is out of scope of the PR or the original usage of this code.

File: earth2mip/lagged_ensembles/core.py (hunk 2 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/core.py:202]: Could you be more specific about what code you see as general purpose here?

File: earth2mip/lagged_ensembles/__main__.py
Comment (nbren12) [reply earth2mip/lagged_ensembles/__main__.py:1]: This PR improves an implementation, I think renaming is a separate PR/issue.

FWIW, I feel the name `python3 -m earth2mip.lagged_ensembles` is relatively informative.

File: earth2mip/initial_conditions/__init__.py [hunk 6]
```diff
@@ -64,16 +89,27 @@ def get_initial_condition_for_model(
...
+    return get_data_from_source(
+        data_source,
+        time,
+        time_loop.in_channel_names,
+        time_loop.grid,
+        time_loop.n_history_levels,
+        time_loop.history_time_step,
+        time_loop.device,
+        time_loop.dtype,
+    )
```
[the remark this answers is not in the document]
Comment (nbren12) [reply earth2mip/initial_conditions/__init__.py:115]: I added this indirection so that I didn't need to touch unrelated parts of the code that rely on `get_initial_condition_for_model` (e.g. ensemble inference). The examples also use `get_initial_condition_for_model`.

If we want to change the public APIs and remove this layer of indirection, we can do so in a separate PR.

File: earth2mip/inference_medium_range.py [hunk 7]
```diff
@@ -148,9 +148,13 @@ def process(initial_time):
 
             lead_time = valid_time - initial_time
             logger.debug(f"{valid_time}")
-            # TODO make this more performant grabs all history steps unnecessarily
-            verification_torch = initial_conditions.get_initial_condition_for_model(
-                time_loop=model, data_source=data_source, time=valid_time
+            verification_torch = initial_conditions.get_data_from_source(
+                data_source=data_source,
+                time=valid_time,
+                channel_names=model.out_channel_names,
```
Comment (yairchn) [inline earth2mip/inference_medium_range.py:154]: should this be `model.in_channel_names`
see #157

File: earth2mip/lagged_ensembles/core.py [hunk 8]
```diff
@@ -13,18 +13,21 @@
...
 async def yield_lagged_ensembles(
     *,
     observations,
     forecast,
-    lags: int = 2,
+    max_lag: int = 2,
+    min_lag: int = -2,
     n: int = 10,
 ):
     """Yield centered lagged ensembles
```
Comment (yairchn) [inline earth2mip/lagged_ensembles/core.py:33]: should we update the text here? not only `centered`

File: earth2mip/initial_conditions/__init__.py [hunk 9]
```diff
@@ -47,13 +55,30 @@ def get_data_source(
...
 
-    dt = time_loop.history_time_step
+    Args:
+        time_levels: the number of time levels to return
+        time_step: the time step of the time levels
+
+    Returns:
+        (time_levels, c, lat, lon) shaped data
+    """
+    dt = time_step
```
[the remark this answers is not in the document]
Comment (nbren12) [reply earth2mip/initial_conditions/__init__.py:79]: It's a worthwhile discussion about data flow, but I think out of scope.  This function is intended to be a private helper that generalizes the `get_initial_condition_for_model`, and is more flexible than that function, and it is needed for graphcast scoring and fixes #156.

Your questions mostly touch more on the general design philosphy of e2mip, and are a bit out of scope for the PR, but in the interest of knowledge transfer I hope these answers are useful:

1. The `.grid` is no longer a schema.Grid Enum. It is an [earth2mip.grid.LatLongGrid](https://github.com/NVIDIA/earth2mip/blob/ebd86637ec853c11e9ff5e717b053351c740e526/earth2mip/grid.py#L25) object, which supports arbitrary lat/lon grids. This was generalization was completed a couple months ago.
2. Time level is roughly described [here](https://github.com/NVIDIA/earth2mip/blob/ebd86637ec853c11e9ff5e717b053351c740e526/earth2mip/time_loop.py#L81), although there is some inconsistent naming.
3. `device`.  Agreed, this and `dtype` probably don't need to be part of this class.

More generally, when fixing bugs are improving performance, I prefer to leave existing APIs/names intact...even if those APIs have some issues...and then fix those APIs in a separate PR.

File: earth2mip/forecasts.py (hunk 1 above)
Comment (nbren12) [reply earth2mip/forecasts.py:91]: I explain the motivation for the `Forecast` API here: https://nvidia.github.io/earth2mip/userguide/concepts.html#forecast

> Many scoring algorithms are most easily expressed as operations over 2D array of states that we call a Forecast Array. The rows of this array correspond to initial times, and the columns to lead times. The size of this array may be unbounded. For example, computing a lead time dependent metrics, such as RMSE corresponds to averaging the square difference of Forecast Arrays of observations and forecasts, and then averaging over the row dimension. This is defined by the [earth2mip.forecasts.Forecast](https://nvidia.github.io/earth2mip/modules/index.html#earth2mip.forecasts.Forecast) interface. Compared to a TimeLoop, a Forecast encapsulates any time handling and initialization logic. One advantage is that an archive of forecasts on disk can be represented as a Forecast (see [earth2mip.forecasts.XarrayForecast](https://nvidia.github.io/earth2mip/modules/index.html#earth2mip.forecasts.XarrayForecast)). This allows using the same code to score both static and streaming forecasts.

File: earth2mip/crps.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+        # c. case x1 < t
+        mask = y > x1
+        val = (x1 - x0) * cdf**2
+        ans += torch.where(mask, val, 0.0)
+
+    # dx [F(x) - H(x-y)]^2 = dx [1 - 0]^2 = dx
+    val = y - ensemble[-1]
+    ans += torch.where(val > 0, val, 0.0)
+    return ans
```
[the remark this answers is not in the document]
Comment (nbren12) [reply earth2mip/crps.py:60]: I hope we can decouple these processes and merge this PR first.

I agree it should be upstreamed, but I think you would be best suited to handle the modulus integration since it sounds like you wanted to refactor tests and compare against the existing implementation.

By merging now, we can ensure that the lagged ensemble scoring works for other people right away and not have to wait for the next modulus release.

Would it help to make this a private API in "earth2mip/lagged_ensembles" for now?

**Comment by nbren12** [comment]
@dallasfoster Thanks for the discussion. I feel it is good knowledge transfer, but perhaps a bit out of scope. Not sure if you have seen these docs: https://nvidia.github.io/earth2mip/userguide/concepts.html#forecast. If not, I highly recommend reading them to understand the current design of e2mip. This PR is however not the place to revisit all aspects of the design.

> A few comments, some of which may turn into issues if desired. I think there is in this PR a pattern of code isolation that should be addressed in order to make the repo more modular.

Yes, this PR is focused on fixing the lagged ensemble script, and some of the pieces may be useful in general, but that is something we can address later I think.

Let's upstream the crps code to modulus. I am okay with duplicating a 1-line helper for weighted averaging a few places in the code.

File: earth2mip/crps.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dallasfoster) [reply earth2mip/crps.py:60]: I can take an AI on upstreaming this code.

File: test/test_crps.py
[the remark this answers is not in the document]
Comment (nbren12) [reply test/test_crps.py:1]: This PR replaces proper-scoring and so I think it makes sense to validate against it. This ensures our old lagged ensemble score are the same as the new ones.

Perhaps we can address when we upstream to modulus.

File: earth2mip/inference_medium_range.py (hunk 7 above)
Comment (nbren12) [reply earth2mip/inference_medium_range.py:154]: No. This just gets verification data. The outputs are compared against verification...not the inputs.

File: earth2mip/initial_conditions/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dallasfoster) [reply earth2mip/initial_conditions/__init__.py:79]: It is not clear that this function is intended to be a private helper. The description and arguments are very general and if I was a user I would not have that context and I might assume that this is intended to be a public function. 

If it is intended to be private then perhaps we should have guidelines for denoting such functions (marking prefix with `_` and clearly denoting it in the function description). 

I think there are two options either we address this in the present PR and clearly mark all of the intended private functions or we make an issue out of it. I will leave the option to you.

For the other functions I agree that we can have a broader discussion about the design of data movement in another forum but sometimes I like having thoughts written for posterity.

File: earth2mip/initial_conditions/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dallasfoster) [reply earth2mip/initial_conditions/__init__.py:115]: Indeed, we will wrap this in the discussion about data movement design.

File: earth2mip/lagged_ensembles/__main__.py
Comment (dallasfoster) [reply earth2mip/lagged_ensembles/__main__.py:1]: I think this comment was more addressing the `Observations` class and `lagged_average_simple` function. Perhaps these two functions can be relocated so that the only method in this file is main?

Comment (dallasfoster) [reply earth2mip/lagged_ensembles/__main__.py:1]: Also, does it make sense for `Observations` to be located in `lagged_ensembles`? This can be an issue and another PR too but I think we should think of another place for such a generic notion.

File: earth2mip/forecasts.py [hunk 11]
```diff
@@ -70,6 +70,25 @@ async def __getitem__(self, i: int):
             yield x
 
 
+class select_channels(Forecast):
```
Comment (NickGeneva) [inline earth2mip/forecasts.py:73]: Odd class name... lower case. If this is kept I think it needs to get changed.

File: earth2mip/lagged_ensembles/core.py (hunk 2 above)
Comment (dallasfoster) [reply earth2mip/lagged_ensembles/core.py:202]: 1. `_convert_ensemble_to_cpu_async` seems like a general purpose utility, not specific to lagged ensemble.
2. `scatter` also seems to be more general purpose utility. As a side note, if this function is meant to be private then we should mark it so, otherwise it would be helpful to have documentation on the inputs.

File: earth2mip/forecasts.py (hunk 11 above)
Comment (NickGeneva) [reply earth2mip/forecasts.py:73]: I dont particularly like this class, feels like a band aid on a bigger problem with the forecast. Granted I dont have a better solution myself for what I understand you want to do.

File: earth2mip/lagged_ensembles/score.py (hunk 3 above)
Comment (dallasfoster) [reply earth2mip/lagged_ensembles/score.py:12]: Since it is such a small line of code, intended to be private, then why does it need to exist as its own function at all? Again, if it is intended to be private then why is that not apparent? 

Can we just drop in `modulus.metrics.climate.reduction.global_mean`?

File: earth2mip/lagged_ensembles/core.py [hunk 12]
```diff
@@ -53,69 +56,50 @@ async def yield_lagged_ensembles(
...
-            if i + n + 1 < nt:
-                obs_buffer.append(await observations[i + n + 1])
-
         i = world_size * i0 + rank
         nsteps = min(nt - world_size * i0 - 1, n)
 
         lead_time = -1
-        async for y in forecast[i]:
+
+        single_forecast = forecast[i] if i < nt else repeat(None)
```
Comment (NickGeneva) [inline earth2mip/lagged_ensembles/core.py:71]: Why not just `yield None` instead this repeat. The present make the iterator infinite when this isn't true. Is that intended?

File: earth2mip/lagged_ensembles/score.py (hunk 5 above)
Comment (dallasfoster) [reply earth2mip/lagged_ensembles/score.py:21]: I agree that it can be out of scope for this PR but we should create an issue to make the scoring function more general.

I can be persuaded that we do actually need multiple score functions like so:
```
earth2mip.lagged_ensemble.score
earth2mip.s2s.score
...
```

File: earth2mip/lagged_ensembles/score.py (hunk 3 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/score.py:12]: It's called a couple of times here, and I didn't want to use copy-paste. Not advertising that it be used everywhere.

I like to see the cos_lat directly in the code, since e.g. graphcast does something different (sin(lat + dlat /2) - sin(lat-dlat/2).

File: earth2mip/forecasts.py (hunk 11 above)
Comment (nbren12) [reply earth2mip/forecasts.py:73]: I made the name lower case so that it looks like a function:
```
regrided_forecast = select_channels(forecast, ["z500"])
scores = some_scoring_workflow(regridded_forecast)
```

A lazy data structure with transformations is a common pattern in data analysis. The idea is that we can transform whole Forecasts objects in a lazy manner similar to xarray.Dataset, a spark RDD, torch DataPipe. 

I agree the Forecast concept may need more thought...it's basically just used by lagged ensembles right now, and I haven't decided whether more workflows should be written using it.

File: earth2mip/lagged_ensembles/core.py (hunk 12 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/core.py:71]: All the ranks need to yield the same number of steps (n) or they will deadlock when the communication happens.

File: earth2mip/lagged_ensembles/__main__.py
Comment (nbren12) [reply earth2mip/lagged_ensembles/__main__.py:1]: Let’s move things in a future PR. This present location of these objects is not something this pr is changing or motivated by.

File: earth2mip/lagged_ensembles/core.py (hunk 2 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/core.py:202]: Let’s defer this. This pr is not defining new general purpose utilities for scattering an ensemble distributed across ranks. Just moving code into helpers as appropriate for improved legibility or the lagged ensemble code.

File: earth2mip/forecasts.py (hunk 11 above)
Comment (nbren12) [reply earth2mip/forecasts.py:73]: @NickGeneva perhaps you can clarify in a separate document what your issues with the forecast object are. This pr does not introduce or change this api, so I think this discussion is out of scope.

File: earth2mip/lagged_ensembles/score.py (hunk 5 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/score.py:21]: resolving, since out of scope.

File: earth2mip/inference_medium_range.py (hunk 7 above)
Comment (yairchn) [reply earth2mip/inference_medium_range.py:154]: in graphcast `model.out_channel_names` includes `tp06` that is not in ERA5 so its not clear to me if this would work

File: earth2mip/forecasts.py (hunk 11 above)
Comment (NickGeneva) [reply earth2mip/forecasts.py:73]: Got it, thanks for clarifying.

File: earth2mip/lagged_ensembles/core.py (hunk 12 above)
Comment (NickGeneva) [reply earth2mip/lagged_ensembles/core.py:71]: Ah, got it. Thanks!

**Comment by NickGeneva** [comment]
@nbren12 looks in general fine to me, although right now I'm not fully familiar with the lagged ensembles. But changes outside seem good enough to get things fixed / discuss more later.

File: earth2mip/inference_medium_range.py (hunk 7 above)
Comment (nbren12) [reply earth2mip/inference_medium_range.py:154]: In practice, graphcast is scored against hdf5 data that have this channel. If the output_channels don't exist in the data that is a bug with the data, not with this routine.

File: earth2mip/initial_conditions/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nbren12) [reply earth2mip/initial_conditions/__init__.py:79]: > If it is intended to be private then perhaps we should have guidelines for denoting such functions 

Agreed, I use the module-level `__all__` for this.

To address your concerns, I have removed this function from that list and added a note to the docstring. I have also renamed the `time_levels` to be more consistent with the `earth2mip.time_loop.TimeLoop` protocol.

device and dtype are still required since it may be beneficial to run the regridding on device.

File: test/test_crps.py
[the remark this answers is not in the document]
Comment (dallasfoster) [reply test/test_crps.py:1]: I agree that we should validate against it, but I think it is also beneficial to have an additional test that doesn't have it as a dependency. In any case, we can discuss further in the upstreaming PR(s). Perhaps the non-dependency tests can live in modulus (which we would rather not have `properscoring` as a dependency) and this test can continue to live here.

File: earth2mip/lagged_ensembles/__main__.py
Comment (dallasfoster) [reply earth2mip/lagged_ensembles/__main__.py:1]: Sure, this can be resolved with a link to an issue.

File: earth2mip/lagged_ensembles/core.py (hunk 2 above)
Comment (dallasfoster) [reply earth2mip/lagged_ensembles/core.py:202]: That's fine, we can resolve with a link to an issue. Perhaps the issue can be broadly centered on distributed utilities.

File: earth2mip/lagged_ensembles/score.py (hunk 3 above)
Comment (dallasfoster) [reply earth2mip/lagged_ensembles/score.py:12]: I still have reservations but I agree that this is low enough priority that we can resolve this without further note.

File: earth2mip/lagged_ensembles/core.py (hunk 2 above)
Comment (nbren12) [reply earth2mip/lagged_ensembles/core.py:202]: on more thought. `scatter` is specific to the distribution pattern used in this particular script, but I can open an issue for the `cpu` async.

Comment (nbren12) [reply earth2mip/lagged_ensembles/core.py:202]: https://github.com/NVIDIA/earth2mip/issues/160

File: earth2mip/lagged_ensembles/__main__.py
Comment (nbren12) [reply earth2mip/lagged_ensembles/__main__.py:1]: https://github.com/NVIDIA/earth2mip/issues/159

File: earth2mip/initial_conditions/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nbren12) [reply earth2mip/initial_conditions/__init__.py:79]: sounds good.

File: earth2mip/forecasts.py (hunk 1 above)
Comment (nbren12) [reply earth2mip/forecasts.py]: I chose to use lower case since this is a `Callable[[Forecast,...], Forecast` that I intended to be used like a function. See https://github.com/NVIDIA/earth2mip/pull/154#discussion_r1440909737 for more info.

I could make it a class `ChannelSubsetForecast` and write a `select_channels(forecast: Forecast, ...) -> ChannelSubsetForecast`, but the effect is the same and this is less work.

**Review by NickGeneva — APPROVED** [review]

**Review by dallasfoster — APPROVED** [review]

**Comment by nbren12** [comment]
The tests are failing when trying to use pytest-regtest: https://github.com/NVIDIA/earth2mip/actions/runs/7493428670/job/20399108089#step:2:1160

I'm not sure why, since the logs show it is being installed. 

@NickGeneva Do you have any ideas about what the issue is?

**Comment by NickGeneva** [comment]
> The tests are failing when trying to use pytest-regtest: https://github.com/NVIDIA/earth2mip/actions/runs/7493428670/job/20399108089#step:2:1160
> 
> I'm not sure why, since the logs show it is being installed.
> 
> @NickGeneva Do you have any ideas about what the issue is?

Not sure, when running locally I had to remove `pytest-regtest` from the TOML to get pytest to function at all. May need to force a lower version to like `1.5.1`. Seems other are having similar issues with the latest release:
https://gitlab.com/uweschmitt/pytest-regtest/-/issues/20

**Comment by nbren12** [comment]
Thanks @NickGeneva for the tip, that fixed it. Merging now.
